src/api/client.ts

```diff
@@ -233,6 +200,144 @@ export class WsClient extends Emittery {
+        assertHolochainSignal(deserializedSignal);
+
+        if (SignalType.System in deserializedSignal) {
+          this.emit("signal", {
+            System: deserializedSignal[SignalType.System],
+          } as Signal);
+        } else {
+          const encodedAppSignal = deserializedSignal[SignalType.App];
+
+          // In order to return readable content to the UI, the signal payload must also be deserialized.
+          const payload = decode(encodedAppSignal.signal);
+
+          const signal: AppSignal = {
+            cell_id: encodedAppSignal.cell_id,
+            zome_name: encodedAppSignal.zome_name,
+            payload,
+          };
+          this.emit("signal", { App: signal } as Signal);
+        }
+      } else if (message.type === "response") {
+        this.handleResponse(message);
+      } else {
+        throw new HolochainError(
+          "UnknownMessageType",
+          `incoming message has unknown type - ${message.type}`
+        );
+      }
+    };
+  }
+
+  private registerCloseListener(socket: IsoWebSocket) {
+    socket.addEventListener(
+      "close",
+      (closeEvent) => {
+        const pendingRequestIds = Object.keys(this.pendingRequests).map((id) =>
+          parseInt(id)
+        );
+        if (pendingRequestIds.length) {
+          pendingRequestIds.forEach((id) => {
+            const error = new HolochainError(
+              "ClientClosedWithPendingRequests",
+              `client closed with pending requests - close event code: ${closeEvent.code}, request id: ${id}`
+            );
+            this.pendingRequests[id].reject(error);
+            delete this.pendingRequests[id];
+          });
+        }
+      },
+      { once: true }
+    );
+  }
+
+  private async reconnectWebsocket(url: URL, token: AppAuthenticationToken) {
+    return new Promise<void>((resolve, reject) => {
+      this.socket = new IsoWebSocket(url, this.options);
+      // This error event never occurs in tests. Could be removed?
+      this.socket.addEventListener(
+        "error",
+        (errorEvent) => {
+          this.authenticationToken = undefined;
+          reject(
+            new HolochainError(
+              "ConnectionError",
+              `could not connect to Holochain Conductor API at ${url} - ${errorEvent.message}`
+            )
+          );
+        },
+        { once: true }
+      );
+
+      const invalidTokenCloseListener = (
+        closeEvent: IsoWebSocket.CloseEvent
+      ) => {
+        this.authenticationToken = undefined;
+        reject(
+          new HolochainError(
+            "InvalidTokenError",
+            `could not connect to ${this.url} due to an invalid app authentication token - close code ${closeEvent.code}`
+          )
+        );
+      };
+      this.socket.addEventListener("close", invalidTokenCloseListener, {
+        once: true,
+      });
+
+      this.socket.addEventListener(
+        "open",
+        async (_) => {
+          const encodedMsg = encode({
```

Comment:
I would have loved to call `authenticate` again here, but that's so convoluted with the `sendHandler` parameter and the `requesterTransformer` and `_request` and `promiseTimeout`, it's head-wrecking. Would be brilliant to simplify that part of the client.